examples/homography_estimation.py

```diff
@@ -0,0 +1,463 @@
+        img2 = img2[0][None, ...]
+        img1_dsts = warp_perspective_norm(H_1_2_mat, img1)
+        path = os.path.join(log_dir, f"{anim_dir}/{it:05d}.png")
+        viz_warp(path, img1[0], img2[0], img1_dsts[0], it, err=err, fc_err=fc_err)
+    anim_path = os.path.join(log_dir, "animation.gif")
+    cmd = f"convert -delay 10 -loop 0 {anim_dir}/*.png {anim_path}"
+    print("Generating gif here: %s" % anim_path)
+    os.system(cmd)
+    shutil.rmtree(anim_dir)
+    return
+
+
+def four_corner_dist(H_1_2, H_1_2_gt, height, width):
+    Hinv_gt = torch.inverse(H_1_2_gt)
+    Hinv = torch.inverse(H_1_2)
+    grid = kornia.utils.create_meshgrid(2, 2, device=Hinv.device)
+    warped_grid = kornia.geometry.transform.homography_warper.warp_grid(grid, Hinv)
+    warped_grid_gt = kornia.geometry.transform.homography_warper.warp_grid(
+        grid, Hinv_gt
+    )
+    warped_grid = (warped_grid + 1) / 2
+    warped_grid_gt = (warped_grid_gt + 1) / 2
+    warped_grid[..., 0] *= width
+    warped_grid[..., 1] *= height
+    warped_grid_gt[..., 0] *= width
+    warped_grid_gt[..., 1] *= height
+    dist = torch.norm(warped_grid - warped_grid_gt, p=2, dim=-1)
+    dist = dist.mean(dim=-1).mean(dim=-1)
+    return dist
+
+
+class SimpleCNN(nn.Module):
+    def __init__(self, D=32):
+        super(SimpleCNN, self).__init__()
+        self.relu = nn.ReLU(inplace=True)
+        self.conv1 = nn.Conv2d(3, D, kernel_size=3, stride=1, padding=1)
+        self.conv2 = nn.Conv2d(D, 3, kernel_size=3, stride=1, padding=1)
+        self.bn1 = nn.BatchNorm2d(D)
+
+    def forward(self, img):
+        x = self.relu(self.bn1(self.conv1(img)))
+        return self.conv2(x)
+
+
+def run():
+    batch_size = 64
+    max_iterations = 50
+    step_size = 0.1
+    verbose = True
+    imgH, imgW = 60, 80
+    use_gpu = True
+    viz_every = 10
+    disp_every = 10
+    save_every = 100
+    num_epochs = 999
+    outer_lr = 1e-4
+    use_cnn = True
+
+    log_dir = os.path.join(os.getcwd(), "viz")
+    os.makedirs(log_dir, exist_ok=True)
+    print("Writing to tensorboard at", os.getcwd())
+    writer = SummaryWriter(log_dir)
+
+    device = "cuda" if torch.cuda.is_available() and use_gpu else "cpu"
+
+    dataset_paths = prepare_data()
+    dataset = HomographyDataset(dataset_paths, imgH, imgW)
+    dataloader = DataLoader(
+        dataset, batch_size=batch_size, shuffle=True, drop_last=True
+    )
+
+    # A simple 2-layer CNN network that maintains the original image size.
+    cnn_model = SimpleCNN()
+    cnn_model.to(device)
+
+    # Set up outer loop optimization.
+    outer_optim = torch.optim.Adam(cnn_model.parameters(), lr=outer_lr)
+
+    itr = 0
+
+    for epoch in range(num_epochs):
+
+        for _, data in enumerate(dataloader):
+
+            outer_optim.zero_grad()
+
+            img1 = data["img1"].to(device)
+            img2 = data["img2"].to(device)
+            Hgt_1_2 = data["H_1_2"].to(device)
+
+            objective = th.Objective()
```

Comment:
Is there a reason why we create the objective and theseus layer inside the loop? Couldn't we create the TheseusLayer outside, around the part where we create the cnn_model, and just populate the data using the forward method?

Reply:
@ddetone @joeaortiz

Reply:
Yes, I agree with Luise, we should create the Theseus layer outside of the loop. I think it was put there before because the loop was over different problem formulations but should definitely be moved outside now.